### PR TITLE
ARROW-8410: [C++] Fix compilation errors on modest ARMv8 platforms (rockpro64, rpi4)

### PR DESCRIPTION
**Pull request by wesm**
These platforms don't support `-march=armv8-a+crc+crypto` so don't think this makes sense to have as a default.

Since ORC has released since a build flag was added to turn off `-Werror` I've upgraded it here and removed the previous warning workarounds

This also resolves ARROW-7968

### CODE REVIEW

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -108,7 +108,10 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
 
   # Arm64 architectures and extensions can lead to exploding combinations.
   # So set it directly through cmake command line.
-  define_option_string(ARROW_ARMV8_ARCH "Arm64 arch and extensions" "armv8-a+crc+crypto")
+  define_option_string(ARROW_ARMV8_ARCH
+                       "Arm64 arch and extensions"
+                       "" # default to None
```
Comment (cyb70289) [inline cpp/cmake_modules/DefineOptions.cmake:113]: We can change the default value to "armv8-a"

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 2]
```diff
@@ -46,7 +46,7 @@ elseif(ARROW_CPU_FLAG STREQUAL "ppc")
   # power compiler flags, gcc/clang only
   set(ARROW_ALTIVEC_FLAG "-maltivec")
   check_cxx_compiler_flag(${ARROW_ALTIVEC_FLAG} CXX_SUPPORTS_ALTIVEC)
-elseif(ARROW_CPU_FLAG STREQUAL "arm")
+elseif(ARROW_CPU_FLAG STREQUAL "arm" AND NOT ("${ARROW_ARMV8_ARCH}" STREQUAL ""))
```
Comment (cyb70289) [inline cpp/cmake_modules/SetupCxxFlags.cmake:49]: then this change is not necessary

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 3]
```diff
@@ -327,7 +328,7 @@ if(ARROW_CPU_FLAG STREQUAL "ppc" AND ARROW_USE_SIMD)
   endif()
 endif()
 
-if(ARROW_CPU_FLAG STREQUAL "arm")
+if(ARROW_CPU_FLAG STREQUAL "arm" AND NOT ("${ARROW_ARMV8_ARCH}" STREQUAL ""))
```
Comment (cyb70289) [inline cpp/cmake_modules/SetupCxxFlags.cmake:331]: and this change not necessary too

**Comment by cyb70289** [comment]
> These platforms don't support `-march=armv8-a+crc+crypto` so don't think this makes sense to have as a default.

Maybe pass `-DARROW_ARMV8_ARCH=armv8-a` when building these platforms and leave the default as is?
True Arm64 servers do support these extensions.

**Comment by wesm** [comment]
> Maybe pass -DARROW_ARMV8_ARCH=armv8-a when building these platforms and leave the default as is?

Having the default build fail on some ARMv8 platforms (on "unserious" ones, I guess =) ) doesn't seem right to me. I would push back against having `-march=avx2` as the default for x64 builds for the same reason. These options are primarily intended for people building packages or developers optimizing for certain CPU capabilities

File: cpp/cmake_modules/DefineOptions.cmake (hunk 1 above)
Comment (wesm) [reply cpp/cmake_modules/DefineOptions.cmake:113]: I'll check this out on my rockpro64 and then make the changes below if all looks good

**Comment by wesm** [comment]
@pitrou thoughts?

**Comment by kszucs** [comment]
@wesm can you test this with docker?

```
ARCH=arm64v8 docker-compose build ubuntu-cpp
ARCH=arm64v8 docker-compose run --rm ubuntu-cpp
```

probably need to add to the volumes section of docker-compose.yml

```yaml
  arm64v8-conda-cache:
  arm64v8-cuda-9.1-cache:
  arm64v8-cuda-10.0-cache:
  arm64v8-cuda-10.1-cache:
  arm64v8-debian-9-cache:
  arm64v8-debian-10-cache:
  arm64v8-fedora-30-cache:
  arm64v8-ubuntu-14.04-cache:
  arm64v8-ubuntu-16.04-cache:
  arm64v8-ubuntu-18.04-cache:
  arm64v8-ubuntu-20.04-cache:
```

**Review by pitrou — APPROVED** [review]
+1

**Comment by wesm** [comment]
@kszucs I'd rather not fiddle more with this, we can tackle Dockerizing the ARM tests in a subsequent PR

**Pull request closed without merging**